mozilla-central-patches/aboutnewtabservice.patch

```diff
@@ -0,0 +1,226 @@
+ 
++const BASE_AS_PATH = "resource://activity-stream/data/content/activity-stream";
+ const ACTIVITY_STREAM_URL = "resource://activity-stream/data/content/activity-stream.html";
+ 
+@@ -24,7 +27,10 @@ const ABOUT_URL = "about:newtab";
+ const IS_MAIN_PROCESS = Services.appinfo.processType === Services.appinfo.PROCESS_TYPE_DEFAULT;
+ 
++const IS_RELEASE_OR_BETA = AppConstants.RELEASE_OR_BETA;
++
+ // Pref that tells if activity stream is enabled
+ const PREF_ACTIVITY_STREAM_ENABLED = "browser.newtabpage.activity-stream.enabled";
+ const PREF_ACTIVITY_STREAM_PRERENDER_ENABLED = "browser.newtabpage.activity-stream.prerender";
++const PREF_ACTIVITY_STREAM_DEBUG = "browser.newtabpage.activity-stream.debug";
+ 
+ function AboutNewTabService() {
+@@ -32,4 +38,9 @@ function AboutNewTabService() {
+   Services.prefs.addObserver(PREF_ACTIVITY_STREAM_ENABLED, this);
+   Services.prefs.addObserver(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED, this);
++  this.initialized = true;
++
++  if (!IS_RELEASE_OR_BETA) {
++    Services.prefs.addObserver(PREF_ACTIVITY_STREAM_DEBUG, this);
++  }
+   this.toggleActivityStream();
+   if (IS_MAIN_PROCESS) {
+@@ -77,4 +88,5 @@ AboutNewTabService.prototype = {
+   _activityStreamEnabled: false,
+   _activityStreamPrerender: false,
++  _activityStreamDebug: false,
+   _overridden: false,
+ 
+@@ -98,10 +110,11 @@ AboutNewTabService.prototype = {
+           this._activityStreamPrerender = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED);
+           Services.obs.notifyObservers(null, "newtab-url-changed", ABOUT_URL);
++        } else if (!IS_RELEASE_OR_BETA && data === PREF_ACTIVITY_STREAM_DEBUG) {
++          this._activityStreamDebug = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_DEBUG, false);
++          Services.obs.notifyObservers(null, "newtab-url-changed", ABOUT_URL);
+         }
+         break;
+       case "quit-application-granted":
+-        Services.obs.removeObserver(this, "quit-application-granted");
+-        Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_ENABLED, this);
+-        Services.prefs.removeObserver(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED, this);
++        this.uninit();
+         if (IS_MAIN_PROCESS) {
+           AboutNewTab.uninit();
+@@ -138,4 +151,7 @@ AboutNewTabService.prototype = {
+     }
+     this._activityStreamPrerender = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_PRERENDER_ENABLED);
++    if (!IS_RELEASE_OR_BETA) {
++      this._activityStreamDebug = Services.prefs.getBoolPref(PREF_ACTIVITY_STREAM_DEBUG, false);
++    }
+     this._newtabURL = ABOUT_URL;
+     return true;
+@@ -152,5 +168,5 @@ AboutNewTabService.prototype = {
+   get defaultURL() {
+     if (this.activityStreamEnabled) {
+-      return this.activityStreamPrerender ? this.activityStreamPrerenderURL : this.activityStreamURL;
++      return `${BASE_AS_PATH}${this.activityStreamPrerender ? "-prerendered" : ""}${this.activityStreamDebug ? "-debug" : ""}.html`
```

Comment:
A comment for the expanded strings (maybe just the file part) would be good for anyone searching for how these files get referenced. E.g., `// activity-stream.html, activity-stream-prerendered.html, activity-stream-debug.html`